File: Python/project-euler/problems/problem_012.py
from utils.code_runner import execute_code


def problem():
    target_divisors = 500
    current_divisors = 0
    current_triangle = 1
    current = 0

    while current_divisors <= target_divisors:
        current += current_triangle
        square_root = 0

        for i in range(1, current + 1):
            square = i * i
            # print(f"i: {i}")
            # print(f"square: {square}")
            # print(f"current: {current}")
            if square > current:
                square_root = i
                # print(f"square_root: {square_root}")
                break

        current_divisors = 1

        for i in range(1, square_root + 1):
            if current % i == 0:
                if current / i == i:
                    current_divisors += 1
                else:
                    current_divisors += 2

        current_triangle += 1

    return current


if __name__ == '__main__':
    execute_code(problem)

File: Java/src/main/java/nl/davydehaas/projecteuler/Problem012.java
package nl.davydehaas.projecteuler;

import static nl.davydehaas.projecteuler.util.SolutionTimer.timeSolution;

import java.util.Arrays;

public class Problem012 {

    void main() {
        timeSolution(Problem012::solve);
    }

    static Number solve() {
        int targetDivisors = 500;
        int currentDivisors = 0;
        int currentTriangle = 1;
        int current = 0;

        while(currentDivisors <= targetDivisors) {
            current += currentTriangle;
            int squareRoot = 0;

            // Calculate the next triangle number.
            for (int i = 1; i < current + 1; i++) {
                int square = i * i;
                if (square > current) {
                    squareRoot = i;
                    break;
                }
            }

            // Calculate the total divisors of the current triangle number.
            currentDivisors = 1;
            for (int i = 1; i < squareRoot + 1; i++) {
                if (current % i == 0) {
                    if (current / i == i) {
                        currentDivisors += 1;
                    } else {
                        currentDivisors += 2;
                    }
                }
            }
            currentTriangle += 1;
        }
        return current;
    }
}
